sdk/python/kfp/dsl/for_loop.py

```diff
@@ -172,26 +175,86 @@ def from_raw_items(
+        return LoopParameterArgument(
             items=raw_items,
             name_code=name_code,
             channel_type=type(raw_items[0]).__name__,
         )
 
-    @classmethod
-    def name_is_loop_argument(cls, name: str) -> bool:
-        """Returns True if the given channel name looks like a loop argument.
 
-        Either it came from a withItems loop item or withParams loop
-        item.
+class LoopArtifactArgument(pipeline_channel.PipelineArtifactChannel):
+    """Represents the artifact arguments that are looped over in a ParallelFor
+    loop.
+
+    The class shouldn't be instantiated by the end user, rather it is
+    created automatically by a ParallelFor ops group.
+
+    To create a LoopArtifactArgument instance, use the factory method::
+
+        LoopArtifactArgument.from_pipeline_channel(...)
+
+
+    Attributes:
+        pipeline_channel: The PipelineArtifactChannel object this
+            LoopArtifactArgument is associated to.
+    """
+
+    def __init__(
+        self,
+        items: pipeline_channel.PipelineArtifactChannel,
+        name_code: Optional[str] = None,
+        name_override: Optional[str] = None,
+        **kwargs,
+    ):
+        """Initializes a LoopArtifactArgument object.
+
+        Args:
+            items: The PipelineArtifactChannel object this LoopArtifactArgument is
+                associated to.
+            name_code: A unique code used to identify these loop arguments.
+                Should match the code for the ParallelFor ops_group which created
+                these LoopArtifactArguments. This prevents parameter name collisions.
+            name_override: The override name for PipelineArtifactChannel.
+            **kwargs: Any other keyword arguments passed down to PipelineArtifactChannel.
         """
-        return  ('-' + cls.LOOP_ITEM_NAME_BASE) in name \
-          or (cls.LOOP_ITEM_PARAM_NAME_BASE + '-') in name
+        if (name_code is None) == (name_override is None):
+            raise ValueError(
+                'Expect one and only one of `name_code` and `name_override` to '
+                'be specified.')
+
+        # We don't support nested lists so `is_artifact_list` is always False.
+        if name_override is None:
+            super().__init__(
+                name=_make_name(name_code), is_artifact_list=False, **kwargs)
+        else:
+            super().__init__(
+                name=name_override, is_artifact_list=False, **kwargs)
+
+        self.items_or_pipeline_channel = items
+        self.is_with_items_loop_argument = not isinstance(
+            items, pipeline_channel.PipelineArtifactChannel)
+
+    @classmethod
+    def from_pipeline_channel(
+        cls,
+        channel: pipeline_channel.PipelineArtifactChannel,
+    ) -> 'LoopArtifactArgument':
+        """Creates a LoopArtifactArgument object from a PipelineArtifactChannel
+        object."""
+        assert channel.is_artifact_list is True, 'Cannot iterate over single Artifacts.'
```

Comment:
Can we raise a formal exception as we do in other cases (as opposed to an `AssertionError`)? And also add a bit more info to help the user associate the error with their LOC? Since the compilation error messages are typically not thrown from the offending LOC in the user's program, it can be hard discern where the error comes from and what to do about it if we don't provide more info.

For example:

```python
if channel.is_artifact_list:
    raise ValueError("Cannot iterate over single a Artifact using 'dsl.ParallelFor'. Expected a list of artifacts as argument to 'items'.")
```

Reply:
Done.